source.go

```diff
@@ -1,98 +1,197 @@
+	}
+	sdk.Logger(ctx).Debug().Msg("Created Pulsar client")
+
+	s.consumer, err = s.client.Subscribe(pulsar.ConsumerOptions{
+		Topic:                       s.config.Topic,
+		SubscriptionName:            s.config.SubscriptionName,
+		Type:                        pulsar.Exclusive,
+		SubscriptionInitialPosition: pulsar.SubscriptionPositionEarliest,
+	})
+	if err != nil {
+		s.client.Close()
+		return fmt.Errorf("failed to create consumer: %w", err)
+	}
+	sdk.Logger(ctx).Debug().Msg("created pulsar consumer")
+
+	if pos != nil {
+		p, err := parsePosition(pos)
+		if err != nil {
+			return err
+		}
+
+		if s.config.SubscriptionName != "" && s.config.SubscriptionName != p.SubscriptionName {
+			return fmt.Errorf("the old position contains a different subscription name than the connector configuration (%q vs %q), please check if the configured subscription name changed since the last run", p.SubscriptionName, s.config.SubscriptionName)
+		}
+
+		s.config.SubscriptionName = p.SubscriptionName
+
+		sdk.Logger(ctx).Info().Str("subscriptionName", s.config.SubscriptionName).Msg("resuming from position")
+	}
+
+	if s.config.SubscriptionName == "" {
+		// this must be the first run of the connector, create a new group ID
+		s.config.SubscriptionName = uuid.NewString()
+		sdk.Logger(ctx).Info().Str("subscriptionName", s.config.SubscriptionName).Msg("assigning source to new subscription")
+	}
+
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	// Read returns a new Record and is supposed to block until there is either
-	// a new record or the context gets cancelled. It can also return the error
-	// ErrBackoffRetry to signal to the SDK it should call Read again with a
-	// backoff retry.
-	// If Read receives a cancelled context or the context is cancelled while
-	// Read is running it must stop retrieving new records from the source
-	// system and start returning records that have already been buffered. If
-	// there are no buffered records left Read must return the context error to
-	// signal a graceful stop. If Read returns ErrBackoffRetry while the context
-	// is cancelled it will also signal that there are no records left and Read
-	// won't be called again.
-	// After Read returns an error the function won't be called again (except if
-	// the error is ErrBackoffRetry, as mentioned above).
-	// Read can be called concurrently with Ack.
-	return sdk.Record{}, nil
+	msg, err := s.consumer.Receive(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed to receive message: %w", err)
+	}
+
+	msgID := msg.ID()
+
+	position := Position{
+		MessageID:        msgID.Serialize(),
+		SubscriptionName: s.config.SubscriptionName,
+	}
+	sdkPos := position.ToSDKPosition()
+
+	metadata := sdk.Metadata{"pulsar.topic": msg.Topic()}
+	metadata.SetCreatedAt(msg.EventTime())
+
+	key := sdk.RawData(msg.Key())
+	payload := sdk.RawData(msg.Payload())
+
+	newRecord := sdk.Util.Source.NewRecordCreate(sdkPos, metadata, key, payload)
```

Comment:
I don't think we need to have all these variables, we can set them here, when creating the record. at least the one-liners

Reply:
fair! I'm more used to the second approach, but ofc it's just a preference